### ISSUE TITLE
NSAutoreleaseNoPool

### ISSUE THREAD

**Issue by glesage** [johnezang/JSONKit#54]
Hi,

I can't figure out if this error is on my side or on the JSONKit's side but I get NSAutoreleaseNoPool errors for a whole bunch of objects:

date - appName[305:7e07] **\* __NSAutoreleaseNoPool(): Object 0x5c590d0 of class __NSDate autoreleased with no pool in place - just leaking
" " Object 0x5c58f70 of class NSCFString                     autoreleased with no pool in place - just leaking
" " Object 0x5c57640 of class __NSCFData                 autoreleased with no pool in place - just leaking
" " Object 0x5c2bc30 of class NSURL                           autoreleased with no pool in place - just leaking
" " Object 0x5c58520 of class __NSArrayM                  autoreleased with no pool in place - just leaking
" " Object 0x5c2cad0 of class NSRecursiveLock          autoreleased with no pool in place - just leaking
" " Object 0x5838ba0 of class NSPathStore2               autoreleased with no pool in place - just leaking
" " Object 0x584fac0 of class __NSArrayI                     autoreleased with no pool in place - just leaking
" " Object 0x5b74120 of class JKSerializer                   autoreleased with no pool in place - just leaking
" " Object 0x5c2cad0 of class NSRecursiveLock          autoreleased with no pool in place - just leaking
" " Object 0x5c53f90 of class __NSCFDictionary          autoreleased with no pool in place - just leaking
" " Object 0x58501c0 of class NSConcreteMutableData autoreleased with no pool in place - just leaking
" " Object 0x5850f60 of class JSONDecoder                autoreleased with no pool in place - just leaking
" " Object 0x5837ee0 of class JKDictionary                    autoreleased with no pool in place - just leaking
" " Object 0x5c5af80 of class JSONDecoder                  autoreleased with no pool in place - just leaking

I get about 100 of each type! Anybody else?

Thanks for the help

**Comment by soffes** [comment]
It looks like you are using JSONKit in a new thread and forgot to setup an autorelease pool in that thread.

**Comment by johnezang** [comment]
This is definitely not a JSONKit issue. You need to follow the Cocoa memory management rules, one of which is that an `NSAutoreleasePool` must always be present.  The most likely cause for your problem is you started a background thread and forgot to create a `NSAutoreleasePool` at the very start of the threads entry function / method.

**Comment by glesage** [comment]
Ok, thanks, I think I found that ASIHTTPRequest uses multiple threads automatically and is causing the problem. However I found that it does initialize an autoreleasePool... weird because nowhere else are there threads in my app...

Thanks for the help

**Comment by j3rm** [comment]
Gefthebest, What was the resolution? I am using ASIHTTPRequest, and I have found that I need to call it after I setup my NSAutoreleasePool, because of Autoreleased objects that just don't go away and cause problems as I am downloading  LARGE amount of data. I found that if I wrap ASI in a NSAutoreleasePool, and then release for each large call, its memory usage is better but the release causes issue with JSONkit over releasing the JKDictionary, or at least I think thats what I found. Thanks for any help.

**Comment by glesage** [comment]
In fact I found that I was doing a "performSelectorInBackground" call which effectively is a change of thread.
I just added NSAutoreleasePool\* pool = [[NSAutoreleasePool alloc] init]; and the [pool release] around the method that I was calling on a separate thread. 
When I say around, I mean inside the method, but must start the pool before any instantiations are made and must release after all releases are done too.
